### ISSUE TITLE
Support extending EventTarget

### ISSUE THREAD

**Issue by zandaqo** [jsdom/jsdom#2156]
### Basic info:
Extending EventTarget doesn't seem to work for me in Jest. Event handlers attached to an instance of an extended EventTarget don't get called when an event is dispatched. Quick debugging points towards [`invokeEventListeners`](https://github.com/jsdom/jsdom/blob/d5639655d32d45b6690c9ac30901136e650eb99a/lib/jsdom/living/events/EventTarget-impl.js#L150) function that returns prematurely due to absence of `_document` on the target.

- **Node.js version:** 8.9.x 
- **jsdom version:** 11.5.1 

### Minimal reproduction case

```js
    it('attaches an event listener', () => {
      class L extends EventTarget {}
      const listener = new L();
      const spy = jest.fn();
      const event = new CustomEvent('a');
      listener.addEventListener('a', spy);
      listener.dispatchEvent(event);
      expect(spy).toBeCalled(); // fails, spy doesn't get called.
    });
```

**Comment by domenic** [comment]
Hmm, you should be getting an error when doing `new L()`; I wonder if something's wrong with our bindings infrastructure...

**Comment by zandaqo** [comment]
Yes, that's what browsers not supporting it do. It would be really nice if we could support it in jsdom though. For now, just adding `_document` pointing to `window.document` seem to make it all work for me.

**Comment by zandaqo** [comment]
On a related note, calling EventTarget's methods (addEventListeners etc) in a context that differs from EventTarget instance doesn't cause error in jsdom, but does so in browsers that implement EventTarget constructor. For example:
```
class L extends EventTarget {}
const l = new L();
const p = new Proxy(l, {});
p.addEventListener('a', () => {});
// ok in jsdom as long as p._document is present
// throws "Illegal invocation" error in Chrome and Firefox.
```

**Comment by huozhi** [comment]
@domenic does jsdom had any plan to implement it?

**Comment by brainkim** [comment]
Hi just to be clear, is the issue jsdom doesn’t implement the EventTarget constructor and I should look elsewhere? Was trying to figure out why custom EventTarget.dispatchEvent stuff just failed silently in some jest tests and found myself here.

**Comment by billyct** [comment]
create a file `mockEventTarget.js`, then copy the code from [Simple implementation of EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget#Simple_implementation_of_EventTarget) and make the `EventTarget` function global

```javascript
// mockEventTarget.js

// copy the code from https://developer.mozilla.org/en-US/docs/Web/API/EventTarget#Simple_implementation_of_EventTarget
var EventTarget = function() {
  this.listeners = {};
};

EventTarget.prototype.listeners = null;
EventTarget.prototype.addEventListener = function(type, callback) {
  if (!(type in this.listeners)) {
    this.listeners[type] = [];
  }
  this.listeners[type].push(callback);
};

EventTarget.prototype.removeEventListener = function(type, callback) {
  if (!(type in this.listeners)) {
    return;
  }
  var stack = this.listeners[type];
  for (var i = 0, l = stack.length; i < l; i++) {
    if (stack[i] === callback){
      stack.splice(i, 1);
      return;
    }
  }
};

EventTarget.prototype.dispatchEvent = function(event) {
  if (!(event.type in this.listeners)) {
    return true;
  }
  var stack = this.listeners[event.type].slice();

  for (var i = 0, l = stack.length; i < l; i++) {
    stack[i].call(this, event);
  }
  return !event.defaultPrevented;
};

// make the EventTarget global
global.EventTarget = EventTarget
```

then config jest's `setupFiles` option with that file

```javascript
// jest.config.js
module.exports = {
  // ... other config
  setupFiles: ['./mockEventTarget.js'],
}
```

it work for me

**Comment by alexkolson** [comment]
This now appears to work (using jsdom@16.7.0), no mocking of `EventTarget` required. Is it still considered a fluke that this works or did something happen to make it work consistently or am I completely off base?

**Comment by espretto** [comment]
This seems to be failing again in `jsdom@19.0.0`. My current `jest.config.js`:
```js
module.exports = {
  setupFiles: ["./jest.setup.js"],
  ...
}
```
and now the mock:
```js
/**
 * jest's jsdom impl fails to call listeners when an event is dispatched.
 * @see https://github.com/jsdom/jsdom/issues/2156
 */
global.EventTarget = class {
  listeners = {};

  addEventListener(type, listener) {
    (this.listeners[type] || (this.listeners[type] = new Set())).add(listener);
  }

  removeEventListener(type, listener) {
    this.listeners[type]?.delete(listener);
  }

  dispatchEvent(event) {
    this.listeners[event.type].forEach(listener => listener(event));
    return !event.defaultPrevented;
  }
};
```

**Comment by vabatta** [comment]
Fails on `jsdom@21.1.2` with default config and can't find a working workaround.

**Comment by aliscie** [comment]
@vabatta did u solve it yet?

**Comment by vabatta** [comment]
@aliscie yes, I used `happy-dom`.

**Comment by domenic** [comment]
This got implemented at some point:

```js
"use strict";
const { JSDOM } = require("jsdom");

const dom = new JSDOM();

class L extends dom.window.EventTarget {}

const listener = new L();
listener.addEventListener("foo", () => {
  console.log("foo");
});
listener.dispatchEvent(new dom.window.Event("foo"));
```

works as expected.

**Comment by isocroft** [comment]
Hello @domenic , I think you might have to reopen this issue:

There seems to be a newer issue around using `EventTarget()` constructor added to **Node.js** since v14.x+

```js
const { JSDOM } = require("jsdom");
const dom = new JSDOM();

const target = new global.EventTarget();
target.dispatchEvent(new dom.window.Event("click")); // throws an error
```

An error is thrown by the above code snippet in all Nodejs versions since v14.x and is further described in detail within [this article](https://www.epicweb.dev/why-i-won-t-use-jsdom)

**Comment by domenic** [comment]
Mixing jsdom and Node.js classes is not and never has been supported. `global.EventTarget` is from Node.js; `dom.window.Event` is from jsdom. Stick to one global, either `dom.window` or `global`.

**Comment by isocroft** [comment]
I am afraid i do not understand your (or the **JSDOM** projects') decision here.

I would graciously ask for a brief explanation.

Are you saying that **JSDOM** does not use the `EventTarget()` constructor natively provided by _Node.js_ v14.x+ ?

Then what does **JSDOM** use under the hood ? A polyfill/shim ? 

This decision is shocking as much as it is surprising

**Comment by Smrtnyk** [comment]
> I am afraid i do not understand your (or the **JSDOM** projects') decision here.
> 
> I would graciously ask for a brief explanation.
> 
> Are you saying that **JSDOM** does not use the `EventTarget()` constructor natively provided by _Node.js_ v14.x+ ?
> 
> Then what does **JSDOM** use under the hood for `dom.window.EventTarget()` ? A polyfill/shim ?
> 
> This decision is shocking as much as it is surprising

Jsdom is a nodejs implementation of a html spec, so why would it use node event target that is not same thing as browser event target?

If you need event target from node natively why are you using jsdom?

**Comment by isocroft** [comment]
Hello @Smrtnyk , I seem ignorant here. Apologies

Please, help me with how the `EventTarget()` API implemented natively in _Node.js_ differs from that implemented in a web browser spec'd by the W3Cs' **DOM Level 2** ?
